radix-engine-stores/src/hash_tree/tree_store.rs

```diff
@@ -41,25 +41,45 @@ pub struct TreeChildEntry {
 
 /// Physical leaf node (which may represent a ReNode or a Substate).
 #[derive(Clone, PartialEq, Eq, Hash, Debug, Categorize, ScryptoEncode, ScryptoDecode)]
-pub struct TreeLeafNode {
-    /// All the remaining nibbles in the _hashed_ `substate_id`.
+pub struct TreeLeafNode<P> {
+    /// All the remaining nibbles in the _hashed_ payload's key.
     pub key_suffix: NibblePath,
-    /// ID of the substate's value in an external storage.
-    pub substate_id: SubstateId,
-    /// An externally-provided hash of the Substate's value.
+    /// Payload; contents depend on the layer.
+    pub payload: P,
+    /// An externally-provided hash of the payload.
     pub value_hash: Hash,
 }
 
+/// Payload of the leafs within the upper (ReNode+Module) layer.
+#[derive(Clone, PartialEq, Eq, Hash, Debug, Categorize, ScryptoEncode, ScryptoDecode)]
+pub struct ReNodeModulePayload {
+    /// ReNode ID.
+    pub re_node_id: RENodeId,
+    /// Module ID.
+    pub node_mode_id: NodeModuleId,
+    /// An embedded root of the descendant Substate layer tree.
+    pub substates_root: TreeNode<SubstateOffset>,
+}
+
+/// "Polymorphic" payload carried by a physical leaf.
+#[derive(Clone, PartialEq, Eq, Hash, Debug, Categorize, ScryptoEncode, ScryptoDecode)]
+pub enum Payload {
+    /// Payload specific to the ReNode+Module layer.
+    ReNodeModule(ReNodeModulePayload),
+    /// Payload specific to the Substate layer.
+    Substate(SubstateOffset),
+}
+
 /// The "read" part of a physical tree node storage SPI.
 pub trait ReadableTreeStore {
     /// Gets node by key, if it exists.
-    fn get_node(&self, key: &NodeKey) -> Option<TreeNode>;
+    fn get_node(&self, key: &NodeKey) -> Option<TreeNode<Payload>>;
```

Comment:
Just to explain what I meant earlier, this could be a `ReadableTreeStore<P: Payload>` and it could return `TreeNode<P>`, and then:

With eg:
```
trait Payload: Clone + PartialEq + Eq + Hash + Debug + Categorize + Encode + Decode {}
```

And eg
```
impl<P: Payload> ReadableTreeStore<P> for X {
...
}
```

Reply:
I see what you mean, but then the `put_at_next_version` would have to take 2 `TreeStore` args in: one for ReNodePayload, one for SubstateOffset. Or a single one with a very weird type constraints? 🤔 Is this what you mean?

Reply:
https://github.com/radixdlt/radixdlt-scrypto/pull/835/files is what I meant :)